versioned/storage/cassandra/src/main/java/org/projectnessie/versioned/storage/cassandra/CassandraBackend.java

```diff
@@ -0,0 +1,454 @@
+            throw new RuntimeException(e);
+          }
+        }
+      }
+
+      // Needed to ensure that writes into the 'result' get populated
+      // TODO verify that this the implementation is correct in terms of the JMM - i.e. that all
+      //  writes into the 'results' array are visible here and to the caller.
+      synchronized (result) {
+        return result;
+      }
+    }
+  }
+
+  PreparedStatement prepared(String cql) {
+    return statements.computeIfAbsent(cql, c -> session.prepare(format(c, keyspace)));
+  }
+
+  BoundStatement buildStatement(String cql, Object[] values) {
+    return prepared(cql).bind(values);
+  }
+
+  boolean executeCas(String cql, Object... values) {
+    try {
+      ResultSet rs = execute(cql, values);
+      return rs.wasApplied();
+    } catch (DriverException e) {
+      handleDriverException(e);
+      return false;
+    }
+  }
+
+  ResultSet execute(String cql, Object... values) {
+    return session.execute(
+        buildStatement(cql, values)
+            .setConsistencyLevel(LOCAL_QUORUM)
+            .setSerialConsistencyLevel(LOCAL_SERIAL));
+  }
+
+  CompletionStage<AsyncResultSet> executeAsync(String cql, Object... values) {
+    return session.executeAsync(
+        buildStatement(cql, values)
+            .setConsistencyLevel(LOCAL_QUORUM)
+            .setSerialConsistencyLevel(LOCAL_SERIAL));
+  }
+
+  void handleDriverException(DriverException e) {
+    if (e instanceof CASWriteUnknownException) {
+      logCASWriteUnknown((CASWriteUnknownException) e);
+    } else {
+      throw e;
+    }
+  }
+
+  @SuppressWarnings("Slf4jDoNotLogMessageOfExceptionExplicitly")
+  private static void logCASWriteUnknown(CASWriteUnknownException e) {
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug(
+          "Assuming CAS failed due to CASWriteUnknownException with message'{}', coordinator: {}, errors: {}, warnings: {}",
+          e.getMessage(),
+          e.getExecutionInfo().getCoordinator(),
+          e.getExecutionInfo().getErrors(),
+          e.getExecutionInfo().getWarnings());
+    }
+  }
+
+  @Override
+  @Nonnull
+  @jakarta.annotation.Nonnull
+  public PersistFactory createFactory() {
+    return new CassandraPersistFactory(this);
+  }
+
+  @Override
+  public void close() {
+    if (closeClient) {
+      session.close();
+    }
+  }
+
+  @Override
+  public void setupSchema() {
+    Metadata metadata = session.getMetadata();
+    Optional<KeyspaceMetadata> keyspace = metadata.getKeyspace(this.keyspace);
+
+    if (!keyspace.isPresent()) {
+      String datacenters =
+          metadata.getNodes().values().stream()
+              .map(Node::getDatacenter)
+              .distinct()
+              .map(dc -> format("'%s': %d", dc, replicationFactor))
+              .collect(Collectors.joining(", "));
+
+      session.execute(
+          format(
+              "CREATE KEYSPACE IF NOT EXISTS %s WITH replication = {'class': 'NetworkTopologyStrategy', %s};",
```

Comment:
This makes me uncomfortable as we are using the same replication factor for all dcs, but it might not be achievable in all of them. If we really need to create the schema on the fly, may I suggest that we require from user to provide the RFs on a per-dc basis.

Reply:
Actually, users should provide an already created keyspace. This one's rather a "damn stupid" default behavior. The Quarkus prod code does (did) require a preexisting keyspace IIRC.

Shall we defer all C* related stuff to later in a follow-up when we re-add C*?

Reply:
I suspect, it was enforced by the C*-quarkus-extension?

Reply:
BUt I've adopted the code to explicitly require the keyspace (and removed the "CREATE KEYSPACE" from prod code)